src/request.rs

```diff
@@ -140,17 +94,176 @@ impl SecretRequest {
-            policies,
-            payload: secret_table,
+        Ok(secret_table)
+    }
+}
+
+trait SecretType {
+    fn payload(&self) -> Result<Vec<u8>>;
+    fn policies(&self) -> Vec<policy::Policy>;
+    fn guid(&self) -> &String;
+}
+
+struct SecretKey {
+    request: RequestDetails,
+}
+
+impl SecretType for SecretKey {
+    fn payload(&self) -> Result<Vec<u8>> {
+        let key = db::get_secret(&self.request.id)?;
+        Ok(match &self.request.format[..] {
+            "binary" => key.into_bytes(),
+            "json" => serde_json::to_string(&key).unwrap().into_bytes(),
+            _ => return Err(anyhow!("Unknown format type")),
         })
     }
 
-    pub fn get_policies(&self) -> &Vec<policy::Policy> {
-        &self.policies
+    fn policies(&self) -> Vec<policy::Policy> {
+        let mut policies = vec![];
+        if let Some(p) = db::get_secret_policy(&self.request.id) {
+            policies.push(p);
+        }
+        policies
+    }
+
+    fn guid(&self) -> &String {
+        &self.request.guid
+    }
+}
+
+#[derive(Serialize, Deserialize, Default)]
+pub struct Key {
+    pub id: String,
+    pub payload: String,
+}
+
+impl Key {
+    pub fn into_bytes(&self) -> Vec<u8> {
+        base64::decode(&self.payload).unwrap()
+    }
+}
+
+struct SecretBundle {
+    request: RequestDetails,
+}
+
+impl SecretType for SecretBundle {
+    fn payload(&self) -> Result<Vec<u8>> {
+        let mut bundle = HashMap::new();
+
+        let secrets = db::get_keyset_ids(&self.request.id)?;
+        for s in secrets {
+            let k = db::get_secret(&s)?;
+            bundle.insert(k.id, k.payload);
+        }
+        Ok(serde_json::to_string(&bundle)?.into_bytes())
+    }
+
+    fn policies(&self) -> Vec<policy::Policy> {
+        let mut policies = vec![];
+
+        if let Some(p) = db::get_keyset_policy(&self.request.id) {
+            policies.push(p);
+        }
+
+        if let std::result::Result::Ok(secrets) = db::get_keyset_ids(&self.request.id) {
+            for s in secrets {
+                if let Some(p) = db::get_secret_policy(&s) {
+                    policies.push(p);
+                }
+            }
+        }
+
+        policies
+    }
+
+    fn guid(&self) -> &String {
+        &self.request.guid
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::db;
+    use crate::grpc::key_broker::RequestDetails;
+
+    #[test]
+    fn test_secret_key() {
+        let secret_id = "keyid-1".to_string();
+        let guid = "2cf13667-ea72-4013-9dd6-155e89c5a28f".to_string();
+        let request = RequestDetails {
+            guid: guid.clone(),
+            format: "binary".to_string(),
+            secret_type: "key".to_string(),
+            id: secret_id.clone(),
+        };
+
+        let secret_value = "dGVzdCBzZWNyZXQ=".to_string(); // "test secret" -> b64
+        let secret_bytes = base64::decode(&secret_value).unwrap();
+        db::insert_secret_only(&secret_id, &secret_value).unwrap();
```

Comment:
consider extracting this to a `setup` function which you will call from several tests. (similarly - extract delete_secret to a `teardown` function). Note that it's not a test framework feature, so `teardown` isn't called when the tests fail.

Reply:
Yeah we might want to adopt some testing framework that supports setup and teardown. Another approach is to have only one function (per file) that is `#[test]` and then have that function call a bunch of different testing routines.

It's not too big of a deal when running via GitHub because the database gets reset each time, but when running locally, you have to manually remove secrets from the database when the tests fail.